[Astronomer][New Version] Added optional renewalMode and endDate(readonly) fields in OfferDetails

vikaschoudharycs097 opened the pull request:
# ARM (Control Plane) API Specification Update Pull Request 

> [!TIP]
> Overwhelmed by all this guidance? See the `Getting help` section at the bottom of this PR description.

## PR review workflow diagram

Please understand this diagram before proceeding. It explains how to get your PR approved & merged.

![diagram](https://github.com/Azure/azure-rest-api-specs/assets/20296335/9e51cff7-28f7-4ae7-99a3-0f4a116c759f)

<details>
<summary>Click here to see the details of Step 1, Breaking Changes review</summary>

If you are in purview of Step 1 of the diagram, follow the [Breaking Changes review process].  
**IMPORTANT!** This applies even if you believe your PR was mislabeled, for any reason, including tool failure.
</details>

<details> 
<summary>Click here to see the details of Step 2, ARM review</summary>

See https://aka.ms/azsdk/pr-arm-review.
</details>

<details>
<summary> Click here to see the diagram footnotes</summary>

### Diagram footnotes

[1] See [ARM review queue] (for **PR merge** queues, see [2]).  
[2] [public repo merge queue], [private repo merge queue] (for **ARM review** queue, [1])  
The ARM reviewer on-call engineer visits the merge queue twice a day, so the approximate ETA for merges is 12 - 24 hours.

</details>

## Purpose of this PR

What's the purpose of this PR? Check the specific option that applies. This is **mandatory**!

  - [ ] New resource provider.
  - [x] New API version for an existing resource provider. (If API spec is not defined in TypeSpec, the PR should have been generated using [OpenAPI Hub](https://aka.ms/openapihub)).
  - [ ] Update existing version for a new feature. (This is applicable only when you are revising a private preview API version.)
  - [ ] Update existing version to fix OpenAPI spec quality issues in S360.
  - [ ] Other, please clarify:
    - _edit this with your clarification_

## Due diligence checklist

To merge this PR, you **must** go through the following checklist and confirm you understood 
and followed the instructions by checking all the boxes:

- [x] I confirm this PR is modifying Azure Resource Manager (ARM) related specifications, and not data plane related specifications.
- [x] I have reviewed following [Resource Provider guidelines](https://aka.ms/rpguidelines), including
  [ARM resource provider contract](https://github.com/Azure/azure-resource-manager-rpc) and
  [REST guidelines](https://github.com/microsoft/api-guidelines/blob/vNext/azure/Guidelines.md) (estimated time: 4 hours).  
  I understand this is required before I can proceed to the diagram Step 2, "ARM API changes review", for this PR.

## Additional information

<details>
<summary> Viewing API changes</summary>

For convenient view of the API changes made by this PR, refer to the URLs provided in the table 
in the `Generated ApiView` comment added to this PR. You can use ApiView to show API versions diff. 

</details>
<details>
<summary>Suppressing failures</summary>

If one or multiple validation error/warning suppression(s) is detected in your PR, please follow the 
[suppressions guide](https://aka.ms/azsdk/pr-suppressions) to get approval.

</details>

## Getting help

- First, please carefully read through this PR description, from top to bottom. Please fill out the `Purpose of this PR` and `Due diligence checklist`.
- To understand what you must do next to merge this PR, see the `Next Steps to Merge` comment. It will appear within few minutes of submitting this PR and will continue to be up-to-date with current PR state.
- For guidance on fixing this PR CI check failures, see the hyperlinks provided in given failure 
  and https://aka.ms/ci-fix.
- For help with PR workflow diagram Step 2 (ARM review), see https://aka.ms/azsdk/pr-arm-review.
- If the PR CI checks appear to be stuck in `queued` state, please add a comment with contents `/azp run`.
  This should result in a new comment denoting a `PR validation pipeline` has started and the checks should be updated after few minutes.
- If the help provided by the previous points is not enough, post to https://aka.ms/azsdk/support/specreview-channel and link to this PR.

[ARM review queue]: https://aka.ms/azsdk/pr-arm-review#about-the-arm-review-queue
[public repo merge queue]: https://github.com/Azure/azure-rest-api-specs/pulls?q=is%3Aopen+is%3Apr+label%3AMergeRequested+draft%3Afalse+sort%3Acreated-asc
[private repo merge queue]: https://github.com/Azure/azure-rest-api-specs-pr/pulls?q=is%3Aopen+is%3Apr+label%3AMergeRequested+-label%3AApproved-OkToMerge+draft%3Afalse+sort%3Acreated-asc
[Breaking Changes review process]: https://eng.ms/docs/cloud-ai-platform/azure-core/azure-core-pm-and-design/trusted-platform-pm-karimb/service-lifecycle-and-actions-team/service-lifecycle-actions-team/apex/media/launchingproductbreakingchanges#breaking-change-process-1

vikaschoudharycs097 committed:
[Astronomer][New Version] Added optional field autoRenew in LiftrBase.OfferDetails

vikaschoudharycs097 committed:
Added new version in readme file

vikaschoudharycs097 committed:
Fixing TypeSpec validation issue

vikaschoudharycs097 committed:
Fixing TypeSpec validation issue

vikaschoudharycs097 committed:
Merge branch 'main' into configurable-auto-renew-feature

vikaschoudharycs097 committed:
Fixed TypeSpec Validation issue

vikaschoudharycs097 committed:
[Fix]: "id" is required to return in response of GET/PUT resource calls

vikaschoudharycs097 committed:
[Fix]: "id" is required to return in response of GET/PUT resource calls

vikaschoudharycs097 committed:
Merge branch 'main' into configurable-auto-renew-feature

ramoka178 commented on specification/liftrastronomer/Astronomer.Astro.Management/main.tsp:
```diff
@@ -32,6 +32,18 @@ enum Versions {
+
+  @doc("Dependent on Azure.ResourceManager.Versions.v1_0_Preview_1, LiftrBase.Versions.v2024_08_27_preview, LiftrBase.Data.Versions.v1_preview")
+  @useDependency(Azure.ResourceManager.Versions.v1_0_Preview_1)
+  @useDependency(LiftrBase.Versions.v2024_08_27_preview)
```
curios. shouldn't these be the stable version instead of preview?

vikaschoudharycs097 replied:
We can add stable version separately and then use it here. But in our case both preview and stable version are exactly same, so we are reusing it instead of defining versions twice (for preview and stable both) at every place.

vikaschoudharycs097 replied:
At the core, we have preview version only (azure-core.tsp).
```
...
enum Versions {
  @doc("Version 1.0-preview.1")
  v1_0_Preview_1: "1.0-preview.1",

  @doc("Version 1.0-preview.2")
  v1_0_Preview_2: "1.0-preview.2",
}
...
```

ramoka178 commented on specification/liftrastronomer/Astronomer.Astro.Management/LiftrBase/main.tsp:
```diff
@@ -65,6 +69,10 @@ model OfferDetails {
+
+  @added(Versions.v2024_08_27_preview)
+  @doc("Subscription auto renewal flag")
+  autoRenew?: boolean;
```
>autoRenew [](http://example.com/codeflow?start=2&length=9)

Standard guidance is: replace boolean/switch properties with a more meaningful enum whenever possible. 

A boolean will forever have two valid values (true or false). A string enum type is always preferred. Also, properties should always provide better values just than True and False. For example two switches "isTypeA" and "isTypeB" should be replaced with one enum "type": [A, B, DefaultType]. Enums are always a more flexible and future proof option because they allow additional values to be added in the future in a non-breaking way, e.g. [Enabled, Disabled, Suspended, Deallocated]. 

Note: do NOT define a 'boolean enum' with two values 'True and False'. This might be easier to 'extend' in terms of types, but semantically its confusing, and no better than a boolean.

vikaschoudharycs097 replied:
As per the requirements from PM, default value of the autoRenew is true. So, we have only two possible values, and I am not seeing any use case where it can have third value. Either autoRenew is enabled or disabled (true or false).

vikaschoudharycs097 replied:
Also, we need to pass this property to marketplace as it is (Marketplace team is expecting true or false for this property).

ramoka178 replied:
You could still accept enums and send the boolean to Saas RP if they need it so. 

We have seen cases where the boolean values are usually not enough as years go by and more values are possible for this state. 
This is a best practice guidance which helps your API to be future proof. This would be good time to make the change, else boolean will keep going on in the next API versions too. 

Please give another thought on this.

vikaschoudharycs097 replied:
Defined enum for auto-renew.

gary-x-li replied:
This is essentially a `boolean enum` (Enable/Disable vs True/False) as mentioned above. How about taking @ramoka178 's suggestion to change it to something like `renewalMode`. It's clearer and more expandable.

> Note: do NOT define a 'boolean enum' with two values 'True and False'. This might be easier to 'extend' in terms of types, but semantically its confusing, and no better than a boolean.

vikaschoudharycs097 replied:
@gary-x-li, are you taking about renaming AutoRenewStatus to RenewalMode or something else?

vikaschoudharycs097 replied:
@gary-x-li, Updated code as per suggestion.

ramoka178 commented on specification/liftrastronomer/Astronomer.Astro.Management/LiftrBase/main.tsp:
```diff
@@ -65,6 +69,10 @@ model OfferDetails {
+
+  @added(Versions.v2024_08_27_preview)
+  @doc("Subscription auto renewal flag")
+  autoRenew?: boolean;
```
>autoRenew [](http://example.com/codeflow?start=2&length=9)

how about a property "renewalMode" and values can be enums of "Auto". "Fixed" or "Default" etc.

vikaschoudharycs097 replied:
We need to pass true/false for this property to Saas RP Endpoint.
![Screenshot 2024-08-28 110513](https://github.com/user-attachments/assets/8cd6d762-00e3-42c8-adb9-89dab3f6f878)

vikaschoudharycs097 committed:
Added "endDate" field in OfferDetails

[Updates]
----------------
- Added "endDate" field in OfferDetails
- Defined AutoRenewStatus enum for autoRenew field
- Defined versioning for stable version "2024-08-27"

weidongxu-microsoft commented:
Java failure due to codegen not supporting extensible enum of non-string type.
```
union AutoRenewStatus {
  integer,

  @doc("Disable/Disabled auto renew")
  Disable: 0,

  @doc("Enable/Enabled auto renew")
  Enable: 1,
}
```

This failure can be ignored.

vikaschoudharycs097 committed:
Updated AutoRenewStatus enum values type to "string"

vikaschoudharycs097 committed:
Merge branch 'main' into configurable-auto-renew-feature

vikaschoudharycs097 committed:
Updated the ordering of AutoRenewStatus enum values

[Update]
----------------
"Enable" is the default value of auto-renew, so placed at first
position.

vikaschoudharycs097 committed:
Merge branch 'Azure:main' into configurable-auto-renew-feature

vikaschoudharycs097 committed:
Merge branch 'configurable-auto-renew-feature' of https://github.com/vikaschoudharycs097/azure-rest-api-specs into configurable-auto-renew-feature

vikaschoudharycs097 committed:
Replaced AutoRenewStatus enum by RenewalMode

vikaschoudharycs097 commented:
/pr RequestMerge

vikaschoudharycs097 committed:
Merge branch 'main' into configurable-auto-renew-feature
